src/unistd.rs

```diff
@@ -2136,3 +2146,424 @@ mod setres {
+    use Result;
+    use errno::Errno;
+    use super::{Error, User, Group, PWGRP_BUFSIZE};
+    use std::{mem, ptr};
+
+    /// Used to get all of the users on the system.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use nix::unistd::Users;
+    /// eprintln!("Users on this system:\nname\tuid");
+    /// Users::new()
+    ///     .map(|e|e.map(
+    ///         |pw| println!("{}\t{}",
+    ///                       pw.name,
+    ///                       pw.uid)))
+    ///     .collect::<Vec<_>>();
+    /// 
+    /// ```
+    ///
+    /// # Safety
+    ///
+    /// This iterator should not be used in different threads without synchronization; while doing so
+    /// will not cause undefined behavior, because modern systems lack re-entrant versions of
+    /// `setpwent` and `endpwent`, it is very likely that iterators running in different threads will
+    /// yield different numbers of items.
+    #[derive(Debug)]
+    pub struct Users(pub usize);
+
+    impl Users {
+        pub fn new() -> Self {
+            unsafe { libc::setpwent(); } 
+            Users(PWGRP_BUFSIZE)
+        }
+
+        pub fn with_bufsize(bufsize: usize) -> Self {
+            unsafe { libc::setpwent(); }
+            Users(bufsize)
+        }
+    }
+
+    impl Iterator for Users {
+        type Item = Result<User>;
+        fn next(&mut self) -> Option<Result<User>> {
+            
+            let mut cbuf = vec![0 as c_char; self.0];
+            let mut pwd: libc::passwd =  unsafe { mem::zeroed() };
+            let mut res = ptr::null_mut();
+            
+            let i = unsafe { 
+                Errno::clear();
+                libc::getpwent_r(&mut pwd, cbuf.as_mut_ptr(), self.0, &mut res) 
+            };
+
+            match i {
+                0 if !res.is_null() => {
+                    Some(Ok(User::from(res)))
+                },
+                libc::ERANGE => { Some(Err(Error::Sys(Errno::last()))) },
+                _ => None
+            }
+        }
+    }
+
+    impl Drop for Users {
+        fn drop(&mut self) {
+            unsafe { libc::endpwent() };
+        }
+    }
+
+    /// Used to get all of the groups on the system.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use nix::unistd::Groups;
+    /// eprintln!("Groups on this system:\nname\tgid");
+    /// Groups::new()
+    ///      .map(|e|e.map(
+    ///          |gr| println!("{}\t{}",
+    ///                        gr.name,
+    ///                        gr.gid)))
+    ///      .collect::<Vec<_>>();
+    /// ```
+    ///
+    /// # Safety
+    ///
+    /// This iterator should not be used in different threads without synchronization; while doing so
+    /// will not cause undefined behavior, because modern systems lack re-entrant versions of
+    /// `setgrent` and `endgrent`, it is very likely that iterators running in different threads will
+    /// yield different numbers of items.
+    #[derive(Debug)]
+    pub struct Groups(pub usize);
```

Comment:
Same as above, can you get away with making the internal value private?

Reply:
done.